tests/cpp/test_topk_device_func.cpp

```diff
@@ -0,0 +1,342 @@
+    tensor = tensor.to(at::kFloat);
+  }
+  auto cpu_tensor = tensor.cpu();
+  auto total_elements = tensor.size(0);
+  return std::vector<DataT>(
+      cpu_tensor.data_ptr<DataT>(),
+      cpu_tensor.data_ptr<DataT>() + total_elements);
+}
+
+// Helper function to validate topk correctness
+template <typename DataT>
+bool validateTopkOrder(
+    const at::Tensor& input_tensor,
+    const at::Tensor& values_tensor,
+    const at::Tensor& indices_tensor,
+    int64_t k,
+    bool largest = true) {
+  NVF_ERROR_EQ(cudaDeviceSynchronize(), cudaSuccess);
+
+  auto input_data = getVector<DataT>(input_tensor);
+  auto output_values = getVector<DataT>(values_tensor);
+  auto output_indices = getVector<int64_t>(indices_tensor);
+
+  // Check that we have k valid results
+  if (static_cast<int64_t>(output_values.size()) < k ||
+      static_cast<int64_t>(output_indices.size()) < k) {
+    return false;
+  }
+
+  // Check valid indices range
+  for (int64_t i = 0; i < k; i++) {
+    if (output_indices[i] < 0 ||
+        output_indices[i] >= static_cast<int64_t>(input_data.size())) {
+      return false;
+    }
+  }
+
+  // Check values match indices
+  for (int64_t i = 0; i < k; i++) {
+    if (output_values[i] != input_data[output_indices[i]]) {
+      return false;
+    }
+  }
+
+  // Check sorting order of the k elements
+  for (int64_t i = 1; i < k; i++) {
+    if (largest) {
+      // For largest, should be in descending order
+      if (output_values[i] > output_values[i - 1]) {
+        return false;
+      }
+    } else {
+      // For smallest, should be in ascending order
+      if (output_values[i] < output_values[i - 1]) {
+        return false;
+      }
+    }
+  }
+
+  // Check that the returned values are actually the true top-k elements
+  // Sort the input data to get the expected top-k values
+  std::vector<DataT> sorted_input = input_data;
+  if (largest) {
+    std::sort(sorted_input.begin(), sorted_input.end(), std::greater<DataT>());
+  } else {
+    std::sort(sorted_input.begin(), sorted_input.end());
+  }
+
+  // Extract the expected top-k values
+  std::vector<DataT> expected_topk(
+      sorted_input.begin(), sorted_input.begin() + k);
+
+  // Extract the actual returned values (first k elements)
+  std::vector<DataT> actual_topk(
+      output_values.begin(), output_values.begin() + k);
+
+  // Compare the expected and actual top-k values
+  if (expected_topk != actual_topk) {
+    return false;
+  }
+
+  return true;
+}
+
+// Basic functionality test
+TEST_F(TopkDeviceFuncTest, BasicTopkFloat) {
+  const int BLOCK_SIZE = 4;
+  const int ITEMS_PER_THREAD = 2;
+  const int total_elements = BLOCK_SIZE * ITEMS_PER_THREAD;
+  const int k = 3;
+
+  auto input_tensor = at::randn(
+      {total_elements},
+      at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0));
+
+  // Note that the output arrays are the same size as the input array
+  // since the CUB-based topk requires the same size of output arrays.
```

Comment:
We shouldn't need to realize that buffer in global memory though. Since the kernel uses registers for cub calls directly.

Reply:
This is just intended to test `blockTopK` as is, which expects each thread has ITEMS_PER_THREAD elements.